src/precice/impl/SolverInterfaceImpl.cpp

```diff
@@ -773,11 +779,54 @@ void SolverInterfaceImpl::setMeshQuad(
     PRECICE_CHECK(mesh->isValidEdgeID(secondEdgeID), "Given EdgeID is invalid!");
     PRECICE_CHECK(mesh->isValidEdgeID(thirdEdgeID), "Given EdgeID is invalid!");
     PRECICE_CHECK(mesh->isValidEdgeID(fourthEdgeID), "Given EdgeID is invalid!");
-    mesh::Edge &e0 = mesh->edges()[firstEdgeID];
-    mesh::Edge &e1 = mesh->edges()[secondEdgeID];
-    mesh::Edge &e2 = mesh->edges()[thirdEdgeID];
-    mesh::Edge &e3 = mesh->edges()[fourthEdgeID];
-    mesh->createQuad(e0, e1, e2, e3);
+    mesh::Edge *e[5];   //Contains the four edges of the quad and the new diagonal that will be created. 
+    e[0] = &mesh->edges()[firstEdgeID];
+    e[1] = &mesh->edges()[secondEdgeID];
+    e[2] = &mesh->edges()[thirdEdgeID];
+    e[3] = &mesh->edges()[fourthEdgeID];
+
+    PRECICE_CHECK(utils::unique_elements(utils::make_array(firstEdgeID, secondEdgeID, thirdEdgeID, fourthEdgeID)),
+                  "The four edge ID's are not unique. Please check that the edges that form the quad are correct.");
+
+    std::array<int,4>  edgeList;  // A list of edge ID numbers
+    edgeList[0] = firstEdgeID;
+    edgeList[1] = secondEdgeID;
+    edgeList[2] = thirdEdgeID;
+    edgeList[3] = fourthEdgeID;
+
+    // Write out a list of all vertex ID's that are eventually connected in 
+    // V0-V1-V2-V3-V0 order to get diagonal distances.
+    mesh::Vertex *vertices[4];
+    // Reorders the edgeList and sets the vertexIDs order.
+    std::array<int,4> vertexIDs = mesh->computeQuadEdgeOrder(edgeList); 
+    
+    // Computes vertex order for the convex quad and reorders the variable vertexIDs
+    PRECICE_CHECK(mesh->computeQuadConvexityFromPoints(vertexIDs),"Quad is not convex."); 
+    vertices[0] = &mesh->vertices()[vertexIDs[0]];
+    vertices[1] = &mesh->vertices()[vertexIDs[1]];
+    vertices[2] = &mesh->vertices()[vertexIDs[2]];
+    vertices[3] = &mesh->vertices()[vertexIDs[3]];
+ 
+    PRECICE_CHECK(utils::unique_elements(utils::make_array(vertices[0]->getCoords(), vertices[1]->getCoords(), 
+                                                           vertices[2]->getCoords(), vertices[3]->getCoords())),
+                  "The four vertices that form the quad are not unique. The resulting shape may be a point, line or triangle." \
+                  "Please check that the edges that form the quad are correct.");
```

Comment:
Shouldn't this check come before the one in line 804?

Reply:
Yes, but then I would have to create the vertices, check for uniqueness, then reorder them and check for convexity and reset the vertices. Perhaps check for uniqueness of the vertexIDs?

Reply:
> Perhaps check for uniqueness of the vertexIDs?

if that covers all case already, good!